src/backend/distributed/commands/role.c

```diff
@@ -0,0 +1,258 @@
+ *
+ * Copyright (c) Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "access/heapam.h"
+#include "access/htup_details.h"
+#if PG_VERSION_NUM >= 120000
+#include "access/table.h"
+#endif
+#include "catalog/catalog.h"
+#include "catalog/pg_authid.h"
+#include "distributed/citus_ruleutils.h"
+#include "distributed/commands.h"
+#include "distributed/commands/utility_hook.h"
+#include "distributed/deparser.h"
+#include "distributed/master_protocol.h"
+#include "distributed/worker_transaction.h"
+#include "nodes/makefuncs.h"
+#include "nodes/parsenodes.h"
+#include "utils/acl.h"
+#include "utils/builtins.h"
+#include "utils/rel.h"
+#include "utils/syscache.h"
+
+static const char * ExtractEncryptedPassword(Oid roleOid);
+static const char * CreateAlterRoleIfExistsCommand(AlterRoleStmt *stmt);
+
+/* controlled via GUC */
+bool EnableAlterRolePropagation = false;
+
+/*
+ * ProcessAlterRoleStmt actually creates the plan we need to execute for alter
+ * role statement.
+ */
+List *
+ProcessAlterRoleStmt(AlterRoleStmt *stmt, const char *queryString)
+{
+	ListCell *optionCell = NULL;
+	List *commands = NIL;
+
+	if (!EnableAlterRolePropagation || !IsCoordinator())
+	{
+		return NIL;
+	}
+
+	foreach(optionCell, stmt->options)
+	{
+		DefElem *option = (DefElem *) lfirst(optionCell);
+
+		if (strcasecmp(option->defname, "password") == 0)
+		{
+			Oid roleOid = get_rolespec_oid(stmt->role, true);
+			const char *encryptedPassword = ExtractEncryptedPassword(roleOid);
+
+			if (encryptedPassword != NULL)
+			{
+				Value *encriptedPasswordValue = makeString((char *) encryptedPassword);
+				option->arg = (Node *) encriptedPasswordValue;
+			}
+			else
+			{
+				option->arg = NULL;
+			}
+
+			break;
+		}
+	}
+	commands = list_make1((void *) CreateAlterRoleIfExistsCommand(stmt));
+
+	return NodeDDLTaskList(ALL_WORKERS, commands);
+}
+
+
+/*
+ * CreateAlterRoleIfExistsCommand creates ALTER ROLE command, from the alter role node
+ *  using the alter_role_if_exists() UDF.
+ */
+static const char *
+CreateAlterRoleIfExistsCommand(AlterRoleStmt *stmt)
+{
+	StringInfoData alterRoleQueryBuffer = { 0 };
+	const char *roleName = RoleSpecString(stmt->role);
+	const char *alterRoleQuery = DeparseTreeNode((Node *) stmt);
+
+	initStringInfo(&alterRoleQueryBuffer);
+	appendStringInfo(&alterRoleQueryBuffer,
+					 "SELECT alter_role_if_exists(%s, %s)",
+					 quote_literal_cstr(roleName),
+					 quote_literal_cstr(alterRoleQuery));
+
+	return alterRoleQueryBuffer.data;
+}
+
+
+/*
+ * ExtractEncryptedPassword extracts the encrypted password of a role. The function
+ * gets the password from the pg_authid table.
+ */
+static const char *
+ExtractEncryptedPassword(Oid roleOid)
+{
+	Relation pgAuthId = heap_open(AuthIdRelationId, AccessShareLock);
+	TupleDesc pgAuthIdDescription = RelationGetDescr(pgAuthId);
+	HeapTuple tuple = SearchSysCache1(AUTHOID, roleOid);
```

Comment:
We can happily pass InvalidOid here as the roleId. So, we should be careful, and return NULL
```C
if (!HeapTupleIsValid(tuple))
{
		retun NULL
}